guardian/fastly-content-fanout
# Create Compute@Edge service to handle websocket and SSE connections

This branch was split to provide a cleaner commit history and PR: see #3 and #4 

## What does this change?

At The Guardian we often report on events that are taking place in real-time, during which circumstances/facts can change very quickly, for example reporting on politics (e.g. general elections) or sports (e.g. Olympics). We also regularly curate fronts to give readers the best coverage of our journalism. In both of these cases it's important to deliver these updates to readers quickly and efficiently, allowing them to stay informed and up-to-date.

Currently our web and app clients poll servers to retrieve updates - this is wasteful in terms of http requests but it can also (depending on the polling frequency) introduce a lag between updates being published and these being available to readers on their devices. Considering this, coupled with effective caching at the CDN, it can be minutes before readers see the latest news.

This PR is a first step towards a different paradigm in delivering the latest news to readers. Instead of clients pulling (via polling) events from the server, the server can push events (when they occur) to clients. If clients can establish long-lived connections to a backend, the backend can broadcast events to all connected clients as they're published. The theory is that this is less wasteful in terms of http requests, but will also lead to a more response and "live" experience for readers.

This PR creates a backend that allows long-lived connections to be established. The backend is a Fastly [Compute@Edge](https://www.fastly.com/products/compute) service that integrates with [fanout](https://docs.fastly.com/products/fanout), allowing [WebSocket](https://en.wikipedia.org/wiki/WebSocket) or [Server-Sent Event](https://en.wikipedia.org/wiki/Server-sent_events) (SSE) connections to be established. This service forms part of an envisaged flow to support delivering real-time events to clients (both web and app):

![compute-at-edge](https://github.com/guardian/fastly-content-fanout/assets/45561419/8a4dfb6d-4c9e-457e-92e2-5558d2f5ad2b)

This PR only supports the connection flow (providing a service clients can communicate with to establish a websockets or SSE connection) but there are companion PRs [here](url) and [here](https://github.com/guardian/facia-tool/pull/1535) to support the server side implementation of the publishing flow.

A critical part of the overall mission in delivering real-time events to clients is understanding what event is relevant to which clients. This can be achieved by the client specifying what "channel" to subscribe to (when establishing the connection). Then, when events are broadcast, sending the relevant event to all clients subscribed to the given channel. From Fastly's [docs](https://developer.fastly.com/learning/concepts/real-time-messaging/fanout/) about using Fanout at the edge:

> When a request is passed to Fanout, it keeps that connection open indefinitely, assigns the connection to one or more channels, and waits for data to be published to those channels. Origin servers or Fastly services may use an API to publish data to a channel, and Fanout will distribute that data to all the clients which have a subscription to that channel.

In the context of real-time events at The Guardian we're currently considering the channel as the content "path". For exmaple, if the reader was viewing the UK home front then the channel would be `/uk`. If they were viewing a politics live blog then the channel would be `/politics/live/2023/dec/20/the-blog`. It is the job of the client to specify the path (= channel) when establishing the connection. The channel is provided as a path parameter in the establishing request.

Given this C@E app is deployed to the Guardian's Fastly account and the service is available under domain `gu-fanout.code.edgecompute.app`, when establishing a websockets connection, the endpoint would be:

`wss://gu-fanout.edgecompute.app/uk` or `wss://gu-fanout.edgecompute.app/politics/live/2023/dec/20/the-blog`

When  establishing a SSE connection, the equivalent endpoints would be:

`https://gu-fanout.edgecompute.app/uk` or `https://gu-fanout.edgecompute.app/politics/live/2023/dec/20/the-blog`

After connections have been established they should remain open for the [default]() period - in the future we may like to consider:
- how to close connections when visiting different fronts/articles
- how to keep the connection open beyond the default idle period (e.g. client sending events?)

Beyond establishing and maintaining long-lived connections, a separate flow will be responsible for broadcasting/publishing events to clients (see [this]() PR).

## How can this change be tested?

To test this change we can establish connections, e.g. from the command line. Using this [utility](https://github.com/websockets/wscat) we can establish a websockets connection:

`wscat -c "wss://gu-fanout.edgecompute.app/uk"`

Where we have specified the path (or channel) as `/uk`. Similarly, we can also establish the equivalent SSE connection:

`curl -i "https://gu-fanout.edgecompute.app/uk"`

When the connection request is made our C@E service hands off the request to Fastly's Fanout service. Fanout manages the connections and makes a subsequent request to our C@E service with a valid [grip](https://grip-framework.github.io/docs/) signature in the header. Given Fanout has established the connection and the C@E service receives the forwarded request (with the expected header) then our service will return a successful response to the client (and we can consider the connection to be open).

To validate that connections have been successfully opened we can send messages to connected clients via a http request, e.g.

`curl -H "Fastly-Key: <fastly-key>" -d '{"items":[{"channel": "uk","formats":{"ws-message":{"content":"hello"},"http-stream":{"content":"hello\n"}}}]}' https://api.fastly.com/service/<fastly-service-id>/publish/`

Where the fastly-key header is a valid API key that's been generated via the Fastly web console and the service ID corresponds to that of the C@E service.

Given the above request is made then all clients connected to the "uk" channel will receive a message of content "hello". It's worth pointing out that, from the above request, we can see that requests to broadcast messages do not go via our C@E service (instead via Fanout).
